File: clients/kotlin/src/test/kotlin/com/microsoft/agenthostprotocol/FileEditTypesTest.kt
package com.microsoft.agenthostprotocol

import com.microsoft.agenthostprotocol.generated.ChangesetFile
import com.microsoft.agenthostprotocol.generated.ChatToolCallReadyAction
import com.microsoft.agenthostprotocol.generated.ContentRef
import com.microsoft.agenthostprotocol.generated.FileEdit
import com.microsoft.agenthostprotocol.generated.FileEditCollection
import com.microsoft.agenthostprotocol.generated.FileEditDiffStats
import com.microsoft.agenthostprotocol.generated.FileEditSide
import com.microsoft.agenthostprotocol.generated.Snapshot
import com.microsoft.agenthostprotocol.generated.ToolCallPendingConfirmationState
import com.microsoft.agenthostprotocol.generated.ToolResultContent
import com.microsoft.agenthostprotocol.generated.ToolResultFileEditContent
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertIs
import kotlin.test.assertNotNull

class FileEditTypesTest {
    private val json: Json = Ahp.json
    private val before = FileEditSide(
        uri = "file:///workspace/old.txt",
        content = ContentRef(
            uri = "ahp-content:/before",
            sizeHint = 32L,
            contentType = "text/plain",
            nonce = "before-v1"
        )
    )
    private val after = FileEditSide(
        uri = "file:///workspace/new.txt",
        content = ContentRef(
            uri = "ahp-content:/after",
            sizeHint = 64L,
            contentType = "text/plain",
            nonce = "after-v2"
        )
    )
    private val edit = FileEdit(
        before = before,
        after = after,
        diff = FileEditDiffStats(added = 2_147_483_648L, removed = 0L)
    )
    private val editWire = """{"before":{"uri":"file:///workspace/old.txt","content":{"uri":"ahp-content:/before","sizeHint":32,"contentType":"text/plain","nonce":"before-v1"}},"after":{"uri":"file:///workspace/new.txt","content":{"uri":"ahp-content:/after","sizeHint":64,"contentType":"text/plain","nonce":"after-v2"}},"diff":{"added":2147483648,"removed":0}}"""
    private val malformedEdits = buildList {
        add("""{"before":{}}""")
        add("""{"after":{"uri":"file:///workspace/file.txt"}}""")
        add("""{"after":{"content":{"uri":"ahp-content:/file"}}}""")
        add("""{"after":{"uri":"file:///workspace/file.txt","content":{}}}""")
        add("""{"after":{"uri":"file:///workspace/file.txt","content":[]}}""")
        add("""{"before":[]}""")
        add("""{"diff":{"added":{}}}""")
        add("""{"diff":{"removed":[]}}""")
    }
    private val malformedCollections = buildList {
        add("{}")
        add("""{"items":null}""")
        add("""{"items":{}}""")
        add("""{"items":[null]}""")
    }

    @Test
    fun `file edit properties have shared serializable types`() {
        val decoded = json.decodeFromString(
            deserializer = FileEdit.serializer(),
            string = editWire
        )
        val decodedSide: FileEditSide = assertNotNull(decoded.after)
        val decodedStats: FileEditDiffStats = assertNotNull(decoded.diff)
        val added: Long? = decodedStats.added
        assertEquals(expected = after, actual = decodedSide)
        assertEquals(expected = before, actual = decoded.before)
        assertEquals(expected = 2_147_483_648L, actual = added)
        assertEquals(expected = edit, actual = decoded)
        assertEquals(
            expected = json.parseToJsonElement(editWire),
            actual = json.parseToJsonElement(
                json.encodeToString(serializer = FileEdit.serializer(), value = edit)
            )
        )
    }

    @Test
    fun `ready actions and pending states expose the same collection type`() {
        val collection = FileEditCollection(items = listOf(edit))
        val ready = json.decodeFromString(
            deserializer = ChatToolCallReadyAction.serializer(),
            string = """{"type":"chat/toolCallReady","turnId":"t1","toolCallId":"tc1","invocationMessage":"Review","edits":{"items":[$editWire]}}"""
        )
        val pending = json.decodeFromString(
            deserializer = ToolCallPendingConfirmationState.serializer(),
            string = """{"status":"pending-confirmation","toolCallId":"tc1","toolName":"edit","displayName":"Edit","invocationMessage":"Review","edits":{"items":[$editWire]}}"""
        )
        val readyEdits: FileEditCollection? = ready.edits
        val pendingEdits: FileEditCollection? = pending.edits
        assertEquals(expected = collection, actual = readyEdits)
        assertEquals(expected = collection, actual = pendingEdits)
    }

    @Test
    fun `file edit tool results expose the same side and statistics types`() {
        val wire = JsonObject(
            json.parseToJsonElement(editWire).jsonObject +
                ("type" to JsonPrimitive("fileEdit"))
        )
        val result = json.decodeFromString(
            deserializer = ToolResultFileEditContent.serializer(),
            string = wire.toString()
        )
        val resultBefore: FileEditSide? = result.before
        val resultAfter: FileEditSide? = result.after
        val resultStats: FileEditDiffStats? = result.diff
        assertEquals(expected = before, actual = resultBefore)
        assertEquals(expected = after, actual = resultAfter)
        assertEquals(expected = edit.diff, actual = resultStats)
    }

    @Test
    fun `empty objects and collections preserve key presence`() {
        assertEquals(
            expected = "{}",
            actual = json.encodeToString(
                serializer = FileEdit.serializer(),
                value = FileEdit()
            )
        )
        assertEquals(
            expected = """{"diff":{}}""",
            actual = json.encodeToString(
                serializer = FileEdit.serializer(),
                value = FileEdit(diff = FileEditDiffStats())
            )
        )
        assertEquals(
            expected = """{"items":[]}""",
            actual = json.encodeToString(
                serializer = FileEditCollection.serializer(),
                value = FileEditCollection(items = emptyList())
            )
        )
    }

    @Test
    fun `malformed known file edits report serialization errors`() {
        for (wire in malformedEdits) {
            assertFailsWith<SerializationException>(message = wire) {
                json.decodeFromString(
                    deserializer = FileEdit.serializer(),
                    string = wire
                )
            }
            val resultWire = JsonObject(
                json.parseToJsonElement(wire).jsonObject +
                    ("type" to JsonPrimitive("fileEdit"))
            )
            assertFailsWith<SerializationException>(message = resultWire.toString()) {
                json.decodeFromString(
                    deserializer = ToolResultContent.serializer(),
                    string = resultWire.toString()
                )
            }
        }
    }

    @Test
    fun `required collection items do not receive silent defaults`() {
        for (wire in malformedCollections) {
            assertFailsWith<SerializationException>(message = wire) {
                json.decodeFromString(
                    deserializer = FileEditCollection.serializer(),
                    string = wire
                )
            }
        }
    }

    @Test
    fun `a malformed file fails the containing changeset snapshot`() {
        val wire = """{"resource":"ahp-changeset:/c1","fromSeq":1,"state":{"status":"ready","files":[{"id":"good","edit":$editWire},{"id":"bad","edit":{"after":{"uri":"file:///workspace/bad.txt","content":{}}}}]}}"""
        assertFailsWith<SerializationException> {
            json.decodeFromString(
                deserializer = Snapshot.serializer(),
                string = wire
            )
        }
    }

    @Test
    fun `known file edit shapes ignore unknown fields`() {
        val wire = """{"futureCollection":1,"items":[{"futureEdit":2,"after":{"uri":"file:///workspace/file.txt","futureSide":3,"content":{"uri":"ahp-content:/file","futureContent":4}},"diff":{"futureStats":5}}]}"""
        val decoded = json.decodeFromString(
            deserializer = FileEditCollection.serializer(),
            string = wire
        )
        assertEquals(
            expected = json.parseToJsonElement("""{"items":[{"after":{"uri":"file:///workspace/file.txt","content":{"uri":"ahp-content:/file"}},"diff":{}}]}"""),
            actual = json.parseToJsonElement(
                json.encodeToString(
                    serializer = FileEditCollection.serializer(),
                    value = decoded
                )
            )
        )
    }

    @Test
    fun `changeset extension metadata remains raw`() {
        val wire = """{"id":"file","edit":$editWire,"_meta":{"vendor":{"keep":true}}}"""
        val decoded = json.decodeFromString(
            deserializer = ChangesetFile.serializer(),
            string = wire
        )
        assertEquals(
            expected = json.parseToJsonElement(wire),
            actual = json.parseToJsonElement(
                json.encodeToString(
                    serializer = ChangesetFile.serializer(),
                    value = decoded
                )
            )
        )
    }

    @Test
    fun `unknown tool result variants retain their raw payload`() {
        val wire = """{"type":"futureFileEdit","payload":{"keep":true}}"""
        val decoded = json.decodeFromString(
            deserializer = ToolResultContent.serializer(),
            string = wire
        )
        val unknown = assertIs<ToolResultContent.Unknown>(decoded)
        assertEquals(expected = json.parseToJsonElement(wire), actual = unknown.raw)
        assertEquals(
            expected = json.parseToJsonElement(wire),
            actual = json.parseToJsonElement(
                json.encodeToString(
                    serializer = ToolResultContent.serializer(),
                    value = decoded
                )
            )
        )
    }
}

File: scripts/file-edit-types.test.ts
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import test from 'node:test';
import { fileURLToPath } from 'node:url';
import { Project } from 'ts-morph';

const root = resolve(dirname(fileURLToPath(import.meta.url)), '..');

function projectForTypes(): Project {
  return new Project({ tsConfigFilePath: resolve(root, 'types/tsconfig.json') });
}

function isJsonObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

function jsonObject(value: unknown): Record<string, unknown> {
  assert.ok(isJsonObject(value), 'expected a JSON object');
  return value;
}

test('file-edit fields reference shared named types', () => {
  const project = projectForTypes();
  const common = project.getSourceFileOrThrow(resolve(root, 'types/common/state.ts'));
  const side = common.getInterfaceOrThrow('FileEditSide');
  const stats = common.getInterfaceOrThrow('FileEditDiffStats');
  const collection = common.getInterfaceOrThrow('FileEditCollection');
  const edit = common.getInterfaceOrThrow('FileEdit');

  assert.deepEqual(side.getProperties().map(p => p.getName()), ['uri', 'content']);
  assert.equal(side.getPropertyOrThrow('uri').getTypeNodeOrThrow().getText(), 'URI');
  assert.equal(side.getPropertyOrThrow('content').getTypeNodeOrThrow().getText(), 'ContentRef');
  assert.ok(side.getProperties().every(p => !p.hasQuestionToken()));
  assert.deepEqual(stats.getProperties().map(p => p.getName()), ['added', 'removed']);
  assert.ok(stats.getProperties().every(p => p.hasQuestionToken()));
  assert.ok(stats.getProperties().every(p => p.getTypeNodeOrThrow().getText() === 'number'));
  assert.equal(collection.getPropertyOrThrow('items').getTypeNodeOrThrow().getText(), 'FileEdit[]');
  assert.equal(collection.getPropertyOrThrow('items').hasQuestionToken(), false);

  for (const name of ['before', 'after']) {
    const property = edit.getPropertyOrThrow(name);
    assert.equal(property.getTypeNodeOrThrow().getText(), 'FileEditSide');
    assert.equal(property.hasQuestionToken(), true);
  }
  assert.equal(edit.getPropertyOrThrow('diff').getTypeNodeOrThrow().getText(), 'FileEditDiffStats');
  assert.equal(edit.getPropertyOrThrow('diff').hasQuestionToken(), true);

  for (const [file, name] of [
    ['types/channels-chat/state.ts', 'ToolCallPendingConfirmationState'],
    ['types/channels-chat/actions.ts', 'ChatToolCallReadyAction'],
  ]) {
    const property = project.getSourceFileOrThrow(resolve(root, file))
      .getInterfaceOrThrow(name).getPropertyOrThrow('edits');
    assert.equal(property.getTypeNodeOrThrow().getText(), 'FileEditCollection');
    assert.equal(property.hasQuestionToken(), true);
  }

  const result = project.getSourceFileOrThrow(resolve(root, 'types/channels-chat/state.ts'))
    .getInterfaceOrThrow('ToolResultFileEditContent');
  assert.ok(result.getExtends().some(base => base.getText() === 'FileEdit'));
});

test('the public TypeScript entry point supports typed file-edit consumers', () => {
  const project = projectForTypes();
  project.createSourceFile(resolve(root, 'types/__file_edit_api_check__.ts'), `
    import type {
      ChatToolCallReadyAction,
      FileEdit,
      FileEditCollection,
      FileEditDiffStats,
      FileEditSide,
      ToolCallPendingConfirmationState,
      ToolResultFileEditContent,
    } from './index.js';

    const side: FileEditSide = {
      uri: 'file:///workspace/file.txt',
      content: { uri: 'ahp-content:/file' },
    };
    const edit: FileEdit = {
      before: side,
      after: side,
      diff: { added: 2147483648, removed: 0 },
    };
    const collection: FileEditCollection = { items: [edit] };
    const ready: ChatToolCallReadyAction['edits'] = collection;
    const pending: ToolCallPendingConfirmationState['edits'] = collection;
    const resultSide: ToolResultFileEditContent['after'] = side;
    const stats: FileEditDiffStats | undefined = edit.diff;
    void [ready, pending, resultSide, stats];
  `);
  assert.deepEqual(
    project.getPreEmitDiagnostics().map(d => d.getMessageText()),
    [],
  );
});

test('named file-edit schemas keep their original required fields and wrappers', () => {
  for (const file of ['state.schema.json', 'actions.schema.json']) {
    const schema = jsonObject(JSON.parse(readFileSync(resolve(root, 'schema', file), 'utf8')));
    const defs = jsonObject(schema.$defs);
    const edit = jsonObject(defs.FileEdit);
    const editProperties = jsonObject(edit.properties);
    assert.equal(jsonObject(editProperties.before).$ref, '#/$defs/FileEditSide');
    assert.equal(jsonObject(editProperties.after).$ref, '#/$defs/FileEditSide');
    assert.equal(jsonObject(editProperties.diff).$ref, '#/$defs/FileEditDiffStats');
    assert.equal(edit.required, undefined);

    const side = jsonObject(defs.FileEditSide);
    assert.deepEqual(side.required, ['uri', 'content']);
    assert.equal(jsonObject(jsonObject(side.properties).content).$ref, '#/$defs/ContentRef');

    const stats = jsonObject(defs.FileEditDiffStats);
    assert.equal(stats.required, undefined);
    assert.equal(jsonObject(jsonObject(stats.properties).added).type, 'number');
    assert.equal(jsonObject(jsonObject(stats.properties).removed).type, 'number');

    const collection = jsonObject(defs.FileEditCollection);
    assert.deepEqual(collection.required, ['items']);
    const items = jsonObject(jsonObject(collection.properties).items);
    assert.equal(items.type, 'array');
    assert.equal(jsonObject(items.items).$ref, '#/$defs/FileEdit');
  }
});
